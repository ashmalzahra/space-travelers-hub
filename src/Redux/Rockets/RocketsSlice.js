import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const url = 'https://api.spacexdata.com/v3/rockets';

const initialState = {
  rockets: [],
  ifSucceed: false,
  isLoading: false,
  error: null,
};

const fetchRockets = createAsyncThunk(
  'rockets/fetchRockets',
  async () => {
    try {
      const { data } = await axios.get(url);
      return data;
    } catch (error) {
      return error.message;
    }
  },
);

const RocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const newArray = state.rockets.map((item) => {
        if (item.rocket_id === action.payload) {
          return { ...item, reserved: true };
        }
        return item;
      });
      return { ...state, rockets: newArray };
    },
    cancelReserveRocket: (state, action) => {
      const newArray = state.rockets.map((item) => {
        if (item.rocket_id === action.payload) {
          return { ...item, reserved: false };
        }
        return item;
      });
      return { ...state, rockets: newArray };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchRockets.fulfilled, (state, action) => ({
        ...state,
        isLoading: false,
        ifSucceed: true,
        rockets: action.payload,
      }))
      .addCase(fetchRockets.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

export const { reserveRocket, cancelReserveRocket } = RocketsSlice.actions;
export { fetchRockets };
export default RocketsSlice.reducer;
