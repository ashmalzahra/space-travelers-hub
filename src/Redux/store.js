import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from './RocketsSlice';
import missionsSlice from './MissionsSlice';
// import missionsReducer from "./MissionsSlice";

const store = configureStore({
  reducer: {
    rockets: rocketsReducer,
    missions: missionsSlice.reducer,
  },
});

export default store;
