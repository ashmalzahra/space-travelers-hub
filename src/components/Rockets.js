import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets } from '../Redux/RocketsSlice';
import Rocket from './Rocket';

const Rockets = () => {
  const dispatch = useDispatch();
  const ifSucceed = useSelector((store) => store.rockets.ifSucceed);
  const isLoading = useSelector((store) => store.rockets.isLoading);
  const rockets = useSelector((store) => store.rockets.rockets);

  useEffect(() => {
    dispatch(fetchRockets());
  }, [ifSucceed, dispatch]);

  let content;
  if (isLoading) {
    content = (
      <div>
        <p>Loading...</p>
      </div>
    );
  } else if (ifSucceed) {
    content = rockets.map((item) => (
      <Rocket
        key={item.rocket_id}
        rocket={item}
      />
    ));
  } else {
    content = (
      <div>
        <p>Something went wrong...</p>
      </div>
    );
  }
  return (
    <div>{content}</div>
  );
};

export default Rockets;
