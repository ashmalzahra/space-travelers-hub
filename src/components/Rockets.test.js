import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import store from '../Redux/store';
import Rockets from './Rockets';

test('Rockets rendered correctly', () => {
  const configureStore = store;
  const component = renderer.create(
    <Provider store={configureStore}>
      <MemoryRouter>
        <Rockets />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// test('Missions renders correctly', () => {
//   const store = configureStore();
//   const component = renderer.create(
//     <Provider store={store}>
//       <MemoryRouter>
//         <Missions />
//       </MemoryRouter>
//     </Provider>,
//   );
//   const tree = component.toJSON();
//   expect(tree).toMatchSnapshot();
// });

// test('Rockets renders correctly', () => {
//   const store = configureStore();
//   const component = renderer.create(
//     <Provider store={store}>
//       <MemoryRouter>
//         <Rockets />
//       </MemoryRouter>
//     </Provider>,
//   );
//   const tree = component.toJSON();
//   expect(tree).toMatchSnapshot();
// });
