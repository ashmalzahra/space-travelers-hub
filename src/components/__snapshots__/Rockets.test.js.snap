// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets rendered correctly 1`] = `
<div>
  <div>
    <p>
      Something went wrong...
    </p>
  </div>
</div>
`;
